firedrake/assemble.py

```diff
@@ -91,21 +92,104 @@ def assemble(expr, tensor=None, bcs=None, form_compiler_parameters=None,
         raise TypeError("Unknown keyword arguments '%s'" % ', '.join(kwargs.keys()))
 
     if isinstance(expr, (ufl.form.Form, slate.TensorBase)):
-        loops = _assemble(expr, tensor=tensor, bcs=solving._extract_bcs(bcs),
-                          form_compiler_parameters=form_compiler_parameters,
-                          mat_type=mat_type,
-                          sub_mat_type=sub_mat_type, appctx=appctx,
-                          diagonal=diagonal,
-                          options_prefix=options_prefix)
-        for l in loops:
-            m = l()
-        return m
+        mat_type, sub_mat_type = get_mat_type(mat_type, sub_mat_type, expr.arguments())
+        assembly_rank = _get_assembly_rank(expr, diagonal)
+        if assembly_rank == AssemblyRank.SCALAR:
+            return _assemble_scalar(expr, tensor, bcs=bcs, form_compiler_parameters=form_compiler_parameters, mat_type=mat_type, sub_mat_type=sub_mat_type, appctx=appctx, options_prefix=options_prefix, assemble_now=assemble_now)
+        elif assembly_rank == AssemblyRank.VECTOR:
+            return _assemble_vector(expr, tensor, bcs=bcs, form_compiler_parameters=form_compiler_parameters, mat_type=mat_type, sub_mat_type=sub_mat_type, diagonal=diagonal, appctx=appctx, options_prefix=options_prefix, assemble_now=assemble_now)
```

Comment:
style: maybe wrap the common arguments into a dict

```python
kwargs = {"bcs": bcs, ...}
if assembly_rank == ...:
    return assemble_scalar(expr, tensor, **kwargs)
elif ...
    return assemble_vector(expr, tensor, diagonal=diagonal, **kwargs)
...
```

Reply:
(My screen is not this wide, or my eyes are not good enough for tiny fonts!)

Reply:
I've created an `_AssemblyOpts` namedtuple to hold all of the options.